modules/users/tests/server/user.server.routes.tests.js

```diff
@@ -262,6 +262,601 @@ describe('User CRUD tests', function () {
+
+    user.save(function (err) {
+      should.not.exist(err);
+      agent.post('/api/auth/forgot')
+        .send({
+          username: ''
+        })
+        .expect(400)
+        .end(function (err, res) {
+          // Handle error
+          if (err) {
+            return done(err);
+          }
+
+          res.body.message.should.equal('Username field must not be blank');
+        });
+
+        agent.post('/api/auth/forgot')
+          .expect(400)
+          .end(function (err, res) {
+            // Handle error
+            if (err) {
+              return done(err);
+            }
+
+            res.body.message.should.equal('Username field must not be blank');
+            return done();
+          });
+
+    });
+  });
+
+  it('forgot password should return 400 for non-local provider set for the user object', function (done) {
+    var provider = 'facebook';
+    user.provider = provider;
+    user.roles = ['user'];
+
+    user.save(function (err) {
+      should.not.exist(err);
+      agent.post('/api/auth/forgot')
+        .send({
+          username: user.username
+        })
+        .expect(400)
+        .end(function (err, res) {
+          // Handle error
+          if (err) {
+            return done(err);
+          }
+
+          res.body.message.should.equal('It seems like you signed up using your ' + user.provider + ' account');
+          return done();
+        });
+    });
+  });
+
+  it('forgot password should return 400 for user password reset request', function (done) {
+    // General comment about this test-
+    // This test practically returns 400 but that's because an actual e-mail will not be
+    // possible to send on most test systems/scenarios.
+    // In this case, we expect 400 response, but we validate that the actual underlying
+    // functionality of the forgot password workflow works as expected, and that the error
+    // is indeed just about failing to send an email.
+
+    user.roles = ['user'];
+
+    user.save(function (err) {
+      should.not.exist(err);
+      agent.post('/api/auth/forgot')
+        .send({
+          username: user.username
+        })
+        .expect(400)
+        .end(function (err, res) {
+          // Handle error
+          if (err) {
+            return done(err);
+          }
+
+          User.findOne({username: user.username.toLowerCase()}, function(err, userRes) {
+            userRes.resetPasswordToken.should.not.be.empty();
+            should.exist(userRes.resetPasswordExpires);
+            res.body.message.should.be.equal('Failure sending email');
+            return done();
+          });
+        });
+    });
+  });
+
+  it('forgot password should be able to reset the password using reset token', function (done) {
+    user.roles = ['user'];
+
+    user.save(function (err) {
+      should.not.exist(err);
+      agent.post('/api/auth/forgot')
+        .send({
+          username: user.username
+        })
+        .expect(400)
```

Comment:
Related to the comment above about e-mails, it's a 400 because it's unable to send e-mails but we later validate and ensure that a reset token has been created.